src/shims/unix/fs.rs

```diff
@@ -1659,6 +1659,68 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
             }
         }
     }
+
+    fn realpath(
+        &mut self,
+        path_op: &OpTy<'tcx, Tag>,
+        processed_path_op: &OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, Pointer<Option<Tag>>> {
+        let this = self.eval_context_mut();
+        this.assert_target_os_is_unix("realpath");
+
+        let pathname = this.read_path_from_c_str(this.read_pointer(path_op)?)?;
+
+        // Reject if isolation is enabled.
+        if let IsolatedOp::Reject(reject_with) = this.machine.isolated_op {
+            this.reject_in_isolation("`realpath`", reject_with)?;
+            let eacc = this.eval_libc("EACCES")?;
+            this.set_last_error(eacc)?;
+            return Ok(Pointer::null());
+        }
+
+        let result = std::fs::canonicalize(pathname);
+        match result {
+            Ok(resolved) => {
+                let processed_ptr = this.read_pointer(processed_path_op)?;
+                let dest = if this.ptr_is_null(processed_ptr)? {
+                    // "If resolved_path is specified as NULL, then realpath() uses
+                    // malloc(3) to allocate a buffer of up to PATH_MAX bytes to hold
+                    // the resolved pathname, and returns a pointer to this buffer.  The
+                    // caller should deallocate this buffer using free(3)."
+                    // <https://man7.org/linux/man-pages/man3/realpath.3.html>
+                    this.alloc_os_str_as_c_str(resolved.as_os_str(), MiriMemoryKind::C.into())?
+                } else {
+                    processed_ptr
+                };
+
+                // Unix MAX_PATH is exposed via libc.
+                #[cfg(not(windows))]
+                let path_max_bytes: u64 = libc::PATH_MAX as u64;
+
+                // Windows MAX_PATH is 260 UTF-16 code units (including the NULL).
+                // Max bytes for code units are 4.
+                // <https://unicode.org/faq/utf_bom.html>
+                #[cfg(windows)]
+                let path_max_bytes = 260 * 4;
+
+                let (wrote_path, _) = this.write_path_to_c_str(&*resolved, dest, path_max_bytes)?;
+
+                if !wrote_path {
+                    // Note that we do not handle `FILENAME_MAX` correctly, but it seems like
```

Comment:
In which sense are we not handling it correctly?

Reply:
Keep in mind I am not an expert and have never delved into this area before . From what I  understand `PATH_MAX` deals with the full string and `FILENAME_MAX` deals with parts between "/". A string such as `[x]/bar` could be under `PATH_MAX` but fail because `[x]` is > `FILENAME_MAX`.

Reply:
Ah this is FILENAME_MAX, nor PATH_MAX, I didn't realize there are two variables here.

I'm not an expert either, but we need to carefully check everything or people will complain that Miri accepts too much or too little code. ;)

Reply:
Ok, this looks like a Linux-ism and even on Linux there are many caveats around `FILENAME_MAX`. I think it's best to just skip over this for now unless people complain.

Reply:
I also don't see FILENAME_MAX even mentioned in the realpath man page.